pkg/shared/utils/arrays/generic_array_utils.go

```diff
@@ -1,12 +1,86 @@
 package arrays
 
+const NotFound = -1
+
+type PredicateFunc[K any] func(x K) bool
+type ReducerFunc[K any, V any] func(accumulatorValue V, currentValue K) V
+type MapperFunc[K any, V any] func(x K) V
+type ConsumerFunc[K any] func(x K)
+
 // FindFirst Finds the first element into the passed in values where the predicate is true
-// WARNING: passed in values are converted to an []interface{} slice and returned value will be an interface{}
-func FindFirst(predicate func(x interface{}) bool, values ...interface{}) (int, interface{}) {
+func FindFirst[K any](values []K, predicate PredicateFunc[K]) (int, K) {
```

Comment:
In Generics and Go the parametrized type is usually denoted as `T`, with the exception on when there's a key and a value (like in the case of the `Map` function). Could we update the occurrences to use `T` to be more idiomatic?